src/awkward/contents/numpyarray.py

```diff
@@ -340,59 +340,34 @@ def _offsets_and_flattened(self, axis, depth):
             )
 
     def _mergeable_next(self, other, mergebool):
-        if isinstance(
-            other,
-            (
-                ak.contents.IndexedArray,
-                ak.contents.IndexedOptionArray,
-                ak.contents.ByteMaskedArray,
-                ak.contents.BitMaskedArray,
-                ak.contents.UnmaskedArray,
-            ),
-        ):
-            return self._mergeable(other._content, mergebool)
+
+        if len(self.shape) > 1:
+            return self._to_regular_primitive()._mergeable(other, mergebool)
```

Comment:
This is the main part of the PR.

In our conversation, I was advocating for

```python
        assert len(self.shape) == 1
```

here, with an upstream conversion of the whole layout to replace `NumpyArray` with `RegularArray` before it enters into `mergeable` _and_ `mergemany`.

I'm assuming that the `NumpyArray` would have to be converted (often a copy) into `RegularArray` again if `mergeable` returns `True` and the `mergemany` step happens. I was just hoping that could be made a single conversion. (It's not a big deal if it isn't.)

Part of that reasoning is that `mergeable`/`mergemany` is an internal API that we have to be careful with, anyway. (Calling `mergemany` on data that would have `mergeable == False` is undefined; I suspect a segfault would result.)

Reply:
We concluded from some experiments (and hindsight) that when `reshape` reduces the number of dimensions, it _does_ have to copy, for any backend.

```python
>>> a = np.arange(3*5*7).reshape(3, 5, 7)
>>> b = a[1:, 1:, 1:]
>>> c = a.T

>>> a[-1, -1, -1] = 123

>>> b
array([[[ 43,  44,  45,  46,  47,  48],
        [ 50,  51,  52,  53,  54,  55],
        [ 57,  58,  59,  60,  61,  62],
        [ 64,  65,  66,  67,  68,  69]],

       [[ 78,  79,  80,  81,  82,  83],
        [ 85,  86,  87,  88,  89,  90],
        [ 92,  93,  94,  95,  96,  97],
        [ 99, 100, 101, 102, 103, 123]]])
>>> c  # also has a "123" in the last element

>>> b_flat = b.reshape((-1,) + b.shape[2:])
>>> c_flat = c.reshape((-1,) + c.shape[2:])

>>> a[-1, -1, -1] = 321

>>> b_flat
array([[ 43,  44,  45,  46,  47,  48],
       [ 50,  51,  52,  53,  54,  55],
       [ 57,  58,  59,  60,  61,  62],
       [ 64,  65,  66,  67,  68,  69],
       [ 78,  79,  80,  81,  82,  83],
       [ 85,  86,  87,  88,  89,  90],
       [ 92,  93,  94,  95,  96,  97],
       [ 99, 100, 101, 102, 103, 123]])
>>> c_flat  # also does not have a "321" in the last element
```

It's because when a dimension is removed, there's nowhere for NumPy to insert the information that there should be a different byte-skip between the affected numbers.

So this `NumpyArray` → `RegularArray` conversion is an $\mathcal{O}(n)$ performance cost for all backends, including NumPy.

JAX had some different performance cost, which NumPy and CuPy didn't have, which I can't remember right now.

Reply:
Okay, yeah, that's cheap-as-free! Now I remember that this, specifically, is the case that JAX doesn't handle as well as NumPy and CuPy. And it's fine to ignore this performance issue in JAX (it's a corner case of a corner case).